runtime/lua/vim/filetype.lua

```diff
@@ -3,21 +3,22 @@ local api = vim.api
 local M = {}
 
 ---@private
-local function starsetf(ft)
+local function starsetf(ft, opts)
   return {
-    function(path)
+    function(path, bufnr)
+      local f = type(ft) == 'function' and ft(path, bufnr) or ft
       if not vim.g.ft_ignore_pat then
-        return ft
+        return f
       end
 
       local re = vim.regex(vim.g.ft_ignore_pat)
-      if re:match_str(path) then
-        return ft
+      if not re:match_str(path) then
```

Comment:
We should probably use `M.matchregex` here instead, but we can do that in a follow up.